ldms/src/core/ldms_heap.h

```diff
@@ -0,0 +1,157 @@
+ *      copyright notice, this list of conditions and the following
+ *      disclaimer in the documentation and/or other materials provided
+ *      with the distribution.
+ *
+ *      Neither the name of Sandia nor the names of any contributors may
+ *      be used to endorse or promote products derived from this software
+ *      without specific prior written permission.
+ *
+ *      Neither the name of Open Grid Computing nor the names of any
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ *      Modified source versions must be plainly marked as such, and
+ *      must not be misrepresented as being the original software.
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __LDMS_HEAP_H__
+#define __LDMS_HEAP_H__
+#include "rrbt.h"
+typedef struct ldms_heap_info {
+	void *start;		/*< The address of the start of the heap */
+	size_t grain;		/*< The minimum allocation size as 2^x */
+	size_t grain_bits;	/*< x in 2^x */
+	size_t size;		/*< The size of the heap in bytes */
+	size_t free_chunks;	/*< number of unallocated chunks current */
+	size_t free_bytes;	/*< number of unallocated grains current */
+	size_t largest;		/*< largest unallocated chunk size in grains */
+	size_t smallest;	/*< smallest unallocated chunk size in grains */
+} *ldms_heap_info_t;
+#define LDMS_HEAP_MIN_SIZE 512
+
+struct ldms_heap {
+	uint32_t grain_bits:8;
+	uint32_t grain:24;	/* Minimum allocation size and alignment */
+	uint64_t size;		/* Size of the heap in bytes */
+	uint32_t gn;            /* Changes when alloc and free  */
+	struct rrbt size_tree;	/* Tree ordered by size */
+	struct rrbt addr_tree;	/* Tree ordered by addr/offset */
+};
+
+typedef struct ldms_heap_instance {
+	struct ldms_heap *data;
+	struct ldms_heap_base *base;
+	pthread_mutex_t lock;
+	rrbt_t size_tree;
+	rrbt_t addr_tree;
+	struct rrbt_instance size_inst;
+	struct rrbt_instance addr_inst;
+} *ldms_heap_t;
+
+struct ldms_heap_base {
+	char signature[8];	/* Identifies start of heap and
+				   prevents 0 from ever being returned
+				   by an allocation */
+	uint8_t start[0];
```

Comment:
should [0] be OVIS_FLEX or is this really *start?

Reply:
It should be OVIS_FLEX. I'm addressing it.